narwhal/executor/src/notifier.rs

```diff
@@ -61,11 +69,9 @@ impl<State: ExecutionState + Send + Sync + 'static> Notifier<State> {
                     )
                     .await;
             }
-            // this is temporary
-            // we will get explicit signal from consensus on where is commit boundary in the future
-            if index.batch_index + 1
-                == index.consensus_output.certificate.header.payload.len() as u64
-            {
+
+            let round = index.consensus_output.certificate.round();
+            if self.committee.leader(round) == self.name {
                 self.callback
                     .notify_commit_boundary(&index.consensus_output)
```

Comment:
I believe this comment now needs to change? https://github.com/MystenLabs/sui/blob/e6f8cee235365d4723e16234e8ff811ea2182e98/narwhal/executor/src/lib.rs#L57

Reply:
I am also a little bit curious why we needed this extra callback here. The way we seem to use this really follows the exact way we execute things. I was thinking whether we could just piggy back the `ConsensusOutput` to mark the commit boundaries. Probably outside of the scope of this PR but since I am looking how this code evolves I would suggest _maybe_ rethinking the approach - unless there is some further design which I ignore here.

Reply:
We do need a callback because sui needs to know when the commit boundary is reached (we actually already have an impl for it in main, so it can provide some context on how it is used).

In terms of how this callback can be implemented I thought about two approaches:

(1) As you say we can perhaps just add a bool to ConsensusOutput (smth like `is_last_in_commit`)
(2) What I thought can be cleaner if we just re-think what consensus output is - instead of sequencing certificates, we sequence Commits. `ConsensusOutput::index` becomes an index of commit rather then height of certificate, `ConsensusOutput::content` becomes `Vec<Certificate>`, instead of `Certificate`. And ExecutionIndexes becomes a 4-touple instead of 3-touple, e.g. holds `(commit_index, certificate_index_in_commit, batch_index_in_certificate, transaction_index_in_batch)`.

I believe that approach (2) is more flexible in terms of what can come next (e.g. for example for MEV would need entire commit in the executor)